### PR TITLE
Allow specifying a subset of parameters to estimate the jacobian by passing the `argnum` kwarg

### LINKED ISSUE #958: Allow tape.jacobian to estimate the jacobian by computing the gradients of a subset of parameters chosen stochastically
Currently, in PennyLane, the `tape.jacobian` method will loop over all parameters in the circuit to construct the Jacobian. If utilizing the parameter-shift rule method of differentiation, for example, this will execute the variational circuit `2*num_parameter` times in order to compute the gradient.

However, a recent result by [Ryan Sweke et al. (2019)](https://arxiv.org/abs/1910.01155) showed that you can randomly sample the parameters to shift at each optimization step. Even in the extreme case where only one parameter is shifted per step, gradient descent convergence remains guaranteed.

This would be a nice feature to add to PennyLane. This could be done as follows:

* Allow a new keyword argument option to be passed to `JacobianTape.jacobian`, for example `k_params` (this is just a placeholder name, final name to be determined).

* Then, modify the internal for loop over all parameters,

  ```python
  for trainable_idx, param_method in enumerate(diff_methods):
  ```

  so that gradient rules are generated by randomly sampling only the specified number of parameters, *if the new keyword argument is provided*:

  ```python
  k_params = options.get("k_params", None)

  params_and_methods = enumerate(diff_methods)
  if stochastic_params is not None:
      params_and_methods = set(random.choices(list(enumerate(diff_methods)), k=k_params))
  
  for trainable_idx, param_method in params_and_methods:
  ```

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
#958 

**Description of the Change:**

Adds the `argnum` keyword argument to the `JacobianTape.jacobian` method inspired by the `argnum` kwarg of the `qml.grad` function. Using this keyword argument the trainable parameters to use for estimating the jacobian can be specified.

The logic is organized into a separate private method to allow easy unit testing and to prepare for further future extensions.

*Note:* as this is an addition to the `JacobianTape`, this option is only applicable for the differentiation methods that depend on such an object and its `jacobian` method: ``"parameter-shift"``, ``"finite-diff"`` and ``"reversible"``.

**Benefits:**

Jacobians can now be estimated by only using a subset of the trainable parameters.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
Closes #958

### CODE REVIEW

File: tests/tape/test_jacobian_tape.py [hunk 1]
```diff
@@ -454,6 +482,55 @@ def test_single_expectation_value(self, tol):
         expected = np.array([[-np.sin(y) * np.sin(x), np.cos(y) * np.cos(x)]])
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
+    def test_single_expectation_value_with_num_params_all(self, tol):
```
Comment (antalszava) [inline tests/tape/test_jacobian_tape.py:485]: This and the following test are using the `test_single_expectation_value` case to make sure that parameter selection works fine.

File: tests/tape/test_jacobian_tape.py [hunk 2]
```diff
@@ -414,6 +414,34 @@ def test_independent_parameters(self):
...
+        num_all_params = len(diff_methods)
+
+        assert all(k in range(num_all_params) for k, _ in res)
+        assert all(v in diff_methods for _, v in res)
+        assert len(res) == num_params if num_params is not None else num_all_params
+
+    @pytest.mark.parametrize("num_params", [0, 2])
+    def test_choose_params_and_methods_warns(self, num_params):
+        """Test that the _choose_params_and_methods helper method warns if too
+        many parameters were specified"""
```
Comment (antalszava) [inline tests/tape/test_jacobian_tape.py:435]:
```suggestion
        many or too few parameters were specified"""
```

**Comment by albi3ro** [comment]
Maybe instead of providing `num_params` to `tape.jacobian`, it accepts a list of `argnum` like `qml.grad`. The choice of `argnum` for the random choice would occur upstream somewhere.  This would make `tape.jacobian` much more versatile and useful as well, providing many more opportunities for speed-up.

**Review by albi3ro — COMMENTED** [review]
Maybe we could instead put this functionality into `_grad_method_validation`? Non-selected parameters simply get their `"grad_method"` set to `0`.  This would keep the logic in `tape.jacobian` simpler for an already complicated function.

**Review by josh146 — APPROVED** [review]
Thanks @antalszava! Great tests, and I'm happy with this PR in it's current form (once @albi3ro's comments are addressed).

@albi3ro makes a good point regarding choosing the random number of parameters higher up in the stack. Currently, the `Jacobian` method computes the derivative based on the `tape.trainable_params`, which is set by the interface. This is worth considering, but should not be a blocker to merging this PR; we can continue to improve this functionality in later PRs.

File: pennylane/tape/jacobian_tape.py [hunk 3]
```diff
@@ -393,6 +395,39 @@ def hessian_pd(self, i, j, params, **options):
         """
         raise NotImplementedError
 
+    @staticmethod
+    def _choose_params_with_methods(diff_methods, num_params):
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/tape/jacobian_tape.py]: Could we do that via `random.seed(my_seed)` in the test? Note that here the `random` library was used for sampling (though open for change if another way or using `numpy` could be preferred).

File: pennylane/tape/jacobian_tape.py [hunk 4]
```diff
@@ -393,6 +395,39 @@ def hessian_pd(self, i, j, params, **options):
...
+        random parameters from the set of trainable parameters.
+
+        Args:
+            diff_methods (list): the ordered list of differentiation methods
+                for each parameter
+            num_params (int or None): The number of parameters to use for
+                computing the jacobian. Specifies the number of parameters to sample.
+
+        Returns:
+            object or list: map of the trainable parameter indices and
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/tape/jacobian_tape.py:418]: Was this a suggestion for the `enumerate` case? Changing `object` to `enumerate` here, though not sure if we also have a `generator` in other cases.

File: .github/CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,9 @@
 
 <h3>New features since last release</h3>
 
+* The number of parameters used to estimate the jacobian can now be specified.
+  [(#1371)](https://github.com/PennyLaneAI/pennylane/pull/1371)
```
[the remark this answers is not in the document]
Comment (antalszava) [reply .github/CHANGELOG.md:6]: Sure! Added

File: pennylane/tape/jacobian_tape.py [hunk 6]
```diff
@@ -393,6 +395,39 @@ def hessian_pd(self, i, j, params, **options):
...
+            differentiation methods
+        """
+        if num_params is None:
+            return enumerate(diff_methods)
+
+        if len(diff_methods) < num_params or num_params < 1:
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/tape/jacobian_tape.py]: Fair point! With the changed logic, the first doesn't make too much sense anymore, as there's a new error now. The second condition is treated separately and returns an empty list.

**Comment by antalszava** [comment]
Thank you for the review comments! :slightly_smiling_face: Not sure, if `_grad_method_validation` would be a good place exactly as that would serve the purpose of validation only, but would be happy to move the logic around in a later addition. For now, might leave it at its current place if that's okay.

**Review by albi3ro — COMMENTED** [review]
Reading this right now, I'm confused about "number of parameters" versus "which parameters".  `argnum` to me connotes the usage in `qml.grad` .  Before I continue to review, I'd like that distinction to be clearer.

**Comment by antalszava** [comment]
Hi @albi3ro, with the latest behaviour of the code, I would think that we have the same behaviour as with `qml.grad`: `argnum` is the argument to specify the indices of the parameters wrt. which we compute/estimate the jacobian. Do we have the same understanding on this one?

Let me know if the confusion comes from the fact that the PR name, changelog, etc. still use the "number of parameters" phrasing (can update those places).

File: .github/CHANGELOG.md [hunk 7]
```diff
@@ -2,6 +2,33 @@
 
 <h3>New features since last release</h3>
 
+* The number of parameters used to estimate the Jacobian can now be specified.
```
Comment (albi3ro) [inline .github/CHANGELOG.md:5]:
```suggestion
* A subset of trainable parameters used to estimate the tape Jacobian can now be specified.
```

File: pennylane/tape/jacobian_tape.py [hunk 8]
```diff
@@ -446,6 +491,10 @@ def jacobian(self, device, params=None, **options):
             order=1 (int): The order of the finite difference method to use. ``1`` corresponds
                 to forward finite differences, ``2`` to centered finite differences.
             shift=pi/2 (float): the size of the shift for two-term parameter-shift gradient computations
+            argnum=None (int, list(int), None): Which argument(s) to compute the jacobian
+                with respect to. When there are fewer parameters specified than the
+                total number of trainable parameters, the jacobian is being estimated by
+                sampling ``argnum`` many parameters from the set of trainable parameters.
```
Comment (albi3ro) [inline pennylane/tape/jacobian_tape.py:497]: Is this sampling thing still the case? Does this need to get updated?

File: tests/tape/test_jacobian_tape.py [hunk 9]
```diff
@@ -471,6 +522,55 @@ def test_single_expectation_value(self, tol):
...
+
+        res = tape.jacobian(dev, argnum=1)  # <--- we only choose one trainable parameter
+        assert res.shape == (1, 2)
+
+        expected = np.array([[-np.sin(y) * np.sin(x), np.cos(y) * np.cos(x)]])
+
+        res = res.flatten()
+        expected = expected.flatten()
+
+        assert any(np.allclose(r, e, atol=tol, rtol=0) for r, e in zip(res, expected))
```
Comment (albi3ro) [inline tests/tape/test_jacobian_tape.py:571]: This is deterministic now, so instead have
```
expected = np.array([[0, np.cos(y)*np.cos(x)]])
```
?

**Review by albi3ro — CHANGES_REQUESTED** [review]
Mostly, the PR just has some hold over's from the previous functionality.

Also, I still think this functionality merges well with `_grad_method_validation`

File: pennylane/tape/jacobian_tape.py [hunk 10]
```diff
@@ -398,6 +400,49 @@ def hessian_pd(self, i, j, params, **options):
...
+            warnings.warn(
+                "No trainable parameters were specified for computing the jacobian.",
+                UserWarning,
+            )
+            return []
+
+        diff_methods_to_sample = map(diff_methods.__getitem__, argnum)
+        mapped_diff_methods = zip(argnum, diff_methods_to_sample)
+
+        return random.sample(list(mapped_diff_methods), k=num_params)
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/tape/jacobian_tape.py:444]: Oh, very good point! Removed the sampling, indeed, was a left-over part from the previous behaviour.

File: pennylane/tape/jacobian_tape.py [hunk 11]
```diff
@@ -398,6 +400,49 @@ def hessian_pd(self, i, j, params, **options):
...
+                with respect to. Specifies the indices of the parameters to sample.
+
+        Returns:
+            enumerate or list: map of the trainable parameter indices and
+            differentiation methods
+        """
+        if argnum is None:
+            return enumerate(diff_methods)
+
+        if not isinstance(argnum, list):
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/tape/jacobian_tape.py:424]: Updated

File: pennylane/tape/jacobian_tape.py [hunk 12]
```diff
@@ -398,6 +400,49 @@ def hessian_pd(self, i, j, params, **options):
...
+            argnum = [argnum]
+
+        if not all(ind in self.trainable_params for ind in argnum):
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/tape/jacobian_tape.py:427]: Good point, after our chat it seems that we'll not consider this (also, seems like `qml.grad` doesn't cover all cases here).

File: pennylane/tape/jacobian_tape.py (hunk 8 above)
Comment (antalszava) [reply pennylane/tape/jacobian_tape.py:497]: Good point, removed that part.

File: tests/tape/test_jacobian_tape.py (hunk 9 above)
Comment (antalszava) [reply tests/tape/test_jacobian_tape.py]: Great point! Updated

**Comment by antalszava** [comment]
@albi3ro thank you for the comments and suggestions here! Great catches. Should have them updated now.

File: pennylane/tape/jacobian_tape.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/tape/jacobian_tape.py:423]: Sure that's updated?

**Review by albi3ro — APPROVED** [review]
Thanks for fixing all the things.

I made a comment about the `if not isinstance(argnum, list):` line, but that's not blocking.

File: pennylane/tape/jacobian_tape.py [hunk 13]
```diff
@@ -398,6 +399,47 @@ def hessian_pd(self, i, j, params, **options):
...
+                compute the Jacobian with respect to.
+
+        Returns:
+            enumerate or list: map of the trainable parameter indices and
+            differentiation methods
+        """
+        if argnum is None:
+            return enumerate(diff_methods)
+
+        if not isinstance(argnum, list):
```
Comment (antalszava) [inline pennylane/tape/jacobian_tape.py:423]:
```suggestion
        if isinstance(argnum, int):
```

**Pull request merged**